File: solutions/ruby/2022/01/lib/solutions.rb
# frozen_string_literal: true

module Solutions202201
  def part_1(input)
    input.split("\n\n").map { |calories_list| calories_list.split("\n").sum(&:to_i) }.max
  end

  def part_2(input)
    input.split("\n\n").map { |calories_list| calories_list.split("\n").sum(&:to_i) }.max(3).sum
  end

  module_function :part_1, :part_2
end

File: solutions/elixir/2022/01/lib/solutions.exs
defmodule Solutions202201 do
  def part_1(input) do
    input
    |> String.trim_trailing()
    |> String.split("\n\n")
    |> Enum.map(fn calories_list ->
      calories_list |> String.split("\n") |> Enum.map(&String.to_integer/1) |> Enum.sum()
    end)
    |> Enum.max()
  end

  def part_2(input) do
    input
    |> String.trim_trailing()
    |> String.split("\n\n")
    |> Enum.map(fn calories_list ->
      calories_list |> String.split("\n") |> Enum.map(&String.to_integer/1) |> Enum.sum()
    end)
    |> Enum.sort(:desc)
    |> Enum.take(3)
    |> Enum.sum()
  end
end
